### PR TITLE
ACLs: Add ACL Server

### LINKED ISSUE #8071: ACLs: Add ACL Auth Server
Add an ACL server to the `contrib/auth/acl` package

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #8071

### Background

Share context and relevant information for the PR: offline discussions, considerations, design decisions etc.

### Bug Fix

If this PR is a bug fix, please let us know about:

1. Problem - The reason for opening the bug
2. Root cause - Discovered root cause after investigation
3. Solution - How the bug was fixed
      
### New Feature

If this PR introduces a new feature, describe it here.

### Testing Details

How were the changes tested?

### Breaking Change?

Does this change break any existing functionality? (API, CLI, Clients)

## Additional info

Logs, outputs, screenshots of changes if applicable (CLI / GUI changes)

### Contact Details

How can we get in touch with you if we need more info? (ex. email@example.com)

### CODE REVIEW

File: .github/workflows/esti.yaml [hunk 1]
```diff
@@ -27,7 +27,10 @@ jobs:
         run: |
           echo "Enable next jobs based on secrets existence: ${{ env.ENABLE_NEXT_JOBS != '' }}"
           echo "secretsavailable=${{ env.ENABLE_NEXT_JOBS != '' }}" >> $GITHUB_OUTPUT
-      
+
```
Comment (arielshaqed) [inline .github/workflows/esti.yaml:30]: So much [nothingness](https://this-page-intentionally-left-blank.org/) (please don't)!

[See 2nd page of this doc](https://ptgmedia.pearsoncmg.com/images/9780132101301/samplepages/0132101300.pdf).

File: api/authorization.yml [hunk 2]
```diff
@@ -313,6 +313,8 @@ components:
           items:
             $ref: "#/components/schemas/Statement"
           minItems: 1
+        acl:
```
Comment (arielshaqed) [inline api/authorization.yml:316]: Can you document this field?  It might mean one of several things.

File: contrib/auth/apigen/oapi-tmpl/chi-interface.tmpl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// chi-interface.tmpl based on https://github.com/deepmap/oapi-codegen/tree/master/pkg/codegen/templates
```
Comment (arielshaqed) [inline contrib/auth/apigen/oapi-tmpl/chi-interface.tmpl:1]: Can we somehow re-use the same templates from lakeFS?  I think it is reasonable for contrib to use main.

File: pkg/auth/service_test.go [hunk 4]
```diff
@@ -1259,11 +1262,13 @@ func TestAPIAuthService_WritePolicy(t *testing.T) {
 			err := s.WritePolicy(ctx, &model.Policy{
 				DisplayName: tt.policyName,
 				CreatedAt:   creationTime,
-				Statement: []model.Statement{{
-					Action:   tt.firstStatementAction,
-					Effect:   tt.firstStatementEffect,
-					Resource: tt.firstStatementResource,
-				}},
+				Statement: []model.Statement{
```
Comment (arielshaqed) [inline pkg/auth/service_test.go:1265]: Please avoid unnecessary whitespace changes.  They bloat a PR that is already too big, and make reviewing harder.  It is perfectly to fine to have a whitespace-only PR before or after a large PR, of course.

File: pkg/auth/setup/setup.go [hunk 5]
```diff
@@ -258,8 +226,12 @@ func CreateInitialAdminUserWithKeys(ctx context.Context, authService auth.Servic
 }
 
 func CreateBaseGroups(ctx context.Context, authService auth.Service, cfg *config.Config, ts time.Time) error {
+	// TODO (niro): need to remove that when transitioning to external auth ACLs server
+	if cfg.IsAuthUISimplified() && cfg.Auth.API.Endpoint != "" {
```
Comment (arielshaqed) [inline pkg/auth/setup/setup.go:230]: I don't understand the 3 arms of logic here.  Can you sketch what the func will look like when the transition finishes?

File: pkg/config/config.go [hunk 6]
```diff
@@ -71,6 +71,77 @@ type S3AuthInfo struct {
 	}
 }
 
+// Database - holds metadata KV configuration
```
Comment (arielshaqed) [inline pkg/config/config.go:74]: This change is unrelated to the PR.  Please revert it and perform is in another PR.

File: pkg/kv/migrations/migrations_test.go [hunk 7]
```diff
@@ -13,7 +13,6 @@ import (
 	authacl "github.com/treeverse/lakefs/contrib/auth/acl"
```
Comment (arielshaqed) [inline pkg/kv/migrations/migrations_test.go:13]: Can we open an issue to get rid of these migrations?  Not sure we need them any longer, and after we clean up auth / ACL I would prefer to lose this code.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

This looks really neat, and having tests makes it _much_ easier to approve.  But please do some manual testing anyway, sorry - it's a large change.

Also a heads-up: this PR includes a tonne of whitespace and formatting changes.  They make it _much harder_ for me to review, meaning it will take less time.  Please revert them.  It would be great to do them in a previous or later PR - jsut not in this one.

File: contrib/auth/apigen/oapi-tmpl/chi-interface.tmpl (hunk 3 above)
Comment (N-o-Z) [reply contrib/auth/apigen/oapi-tmpl/chi-interface.tmpl]: Trying

File: pkg/config/config.go (hunk 6 above)
Comment (N-o-Z) [reply pkg/config/config.go:74]: Of course it is related to this PR.
I'm reusing the Database struct for the ACL server config.
If you want me to break this into pieces and first merge this change I can do that. But this is required for this PR

File: pkg/auth/setup/setup.go (hunk 5 above)
Comment (N-o-Z) [reply pkg/auth/setup/setup.go:230]: When the transition finishes there will be no more call to CreateACLBaseGroups.
This part is done on the ACL server side (whether as part of a first `run` or an explicit `setup` command is still TBD.
For RBAC it will still create the base groups via lakeFS.

File: pkg/kv/migrations/migrations_test.go (hunk 7 above)
Comment (N-o-Z) [reply pkg/kv/migrations/migrations_test.go:13]: https://github.com/treeverse/lakeFS/issues/8077 FYI @guy-har

**Comment by N-o-Z** [comment]
> Thanks!
> 
> This looks really neat, and having tests makes it _much_ easier to approve. But please do some manual testing anyway, sorry - it's a large change.
> 
> Also a heads-up: this PR includes a tonne of whitespace and formatting changes. They make it _much harder_ for me to review, meaning it will take less time. Please revert them. It would be great to do them in a previous or later PR - jsut not in this one.

Thanks for the review @arielshaqed 

Of course the changes have also been tested manually in addition to the esti job (it would have taken me forever to try and create the esti job to actually work otherwise).
Regarding the whitespace remark, I looked over the PR and found 2 places where this occurred 1 on esti.yaml and the other was a small line fix. I'm not sure this justifies reverting it - the way these changes are reflected in github is pretty contained.
I understand this is a big PR (I didn't plan on it but I couldn't also open the PR without proving this actually works i.e. esti tests), so I understand if this might take a while to review! 🙏🏽

File: .github/workflows/esti.yaml (hunk 1 above)
Comment (N-o-Z) [reply .github/workflows/esti.yaml]: oops 😅

File: contrib/auth/acl/controller.go [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+	writeError(w, http.StatusNotImplemented, "Not implemented")
+}
+
+func (c *Controller) AttachPolicyToUser(w http.ResponseWriter, _ *http.Request, _, _ string) {
+	writeError(w, http.StatusNotImplemented, "Not implemented")
+}
+
+func (c *Controller) GetVersion(w http.ResponseWriter, _ *http.Request) {
+	writeResponse(w, http.StatusOK, apigen.VersionConfig{
+		Version: "dev",
```
Comment (guy-har) [inline contrib/auth/acl/controller.go:610]: ?

**Review by guy-har — COMMENTED** [review]
BIG change! 
Nice work, added some comments, most are questions or nit, my only concern is regarding the int64 ids

File: pkg/api/controller.go [hunk 9]
```diff
@@ -840,7 +840,7 @@ func (c *Controller) ListGroups(w http.ResponseWriter, r *http.Request, params a
 	ctx := r.Context()
 	c.LogAction(ctx, "list_groups", r, "", "", "")
 	groups, paginator, err := c.Auth.ListGroups(ctx, &model.PaginationParams{
-		After:  paginationAfter(params.After),
+		After:  PaginationAfter(params.After),
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/api/controller.go:843]: Reverted!

File: contrib/auth/acl/controller.go [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+	if c.handleAPIError(w, err) {
+		return
+	}
+	response := apigen.CredentialsWithSecret{
+		UserName:        swag.String(credentials.Username),
+		AccessKeyId:     credentials.AccessKeyID,
+		SecretAccessKey: credentials.SecretAccessKey,
+		CreationDate:    credentials.IssuedDate.Unix(),
+	}
+	response.UserId, _ = strconv.ParseInt(credentials.Username, 10, 64)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply contrib/auth/acl/controller.go:51]: Removed

File: contrib/auth/acl/controller.go [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+	// if external user management is enabled
+	// this is to mitigate this issue: https://github.com/treeverse/cloud-controlplane/issues/1281
+	ctx := r.Context()
+	u := &model.User{
+		CreatedAt:    time.Now().UTC(),
+		Username:     body.Username,
+		FriendlyName: body.FriendlyName,
+		Email:        body.Email,
+		Source:       swag.StringValue(body.Source),
+		ExternalID:   body.ExternalId,
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply contrib/auth/acl/controller.go:358]: Removed it

File: contrib/auth/acl/controller.go (hunk 8 above)
Comment (N-o-Z) [reply contrib/auth/acl/controller.go:610]: It's hardcoded version of the server - we don't really need to maintain versioning in this example package

File: contrib/auth/acl/controller.go [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+func writeError(w http.ResponseWriter, code int, v interface{}) {
+	apiErr := apigen.Error{
+		Message: fmt.Sprint(v),
+	}
+	writeResponse(w, code, apiErr)
+}
+
+func writeResponse(w http.ResponseWriter, code int, response interface{}) {
+	log := logging.ContextUnavailable()
+	if response == nil {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply contrib/auth/acl/controller.go:638]: It's a simplified version of the response handler

File: contrib/auth/acl/cmd/acl/cmd/root.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/treeverse/lakefs/pkg/config"
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+var cfgFile string
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:     "acl",
+	Short:   "acl is an acl auth server example for LakeFS",
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply contrib/auth/acl/cmd/acl/cmd/root.go:24]: I think it's better to leave it, as this is a reference server and not something that should be used in production or provides any guarantees

**Review by guy-har — APPROVED** [review]
LGTM
